doc: add "Stored and View Schema" doc for tree

CraigMacomber opened the pull request:
Document for #9234.

Commit:
doc: add "Container and Application Schema" doc for tree

Commit:
fix value schema, and add note about makeAnyType

Commit:
capatalize Fluid

CraigMacomber commented on packages/dds/tree/docs/wip/Container and Application Schema.md:
```diff
@@ -0,0 +1,194 @@
+# Definitions
+
+-   `container data` : data stored in the Fluid container by the shared tree dds.
+-   `container schema` : a set of constraints it is valid to assume the `container data` meets, and that must be maintained when editing (including through conflicts)
```
The main distinction here is where its stored (in the container/dds/document with the tree) instead of in the application (its source tree and some runtime data).

I don't like "root schema" since the application root schema.
Some alternative names to replace container vs application:

For "container":

1. document
2. stored/storage
4. recorded/recording
5. persisted/persistent

For application:
1. view
2. projection

I think my new favorite is: stored schema and view schema.

Since this document is mostly not fluid specific (its mostly conceptual/general), I could rename "container data" to "stored data" (which consists of a "stored document" + its "stored schema").

Does that seem like a good idea?

jack-williams replied:
IMO document schema matches most closely, based on how I read the proposal. A node (or tree or document) must have a type, and that type is agreed upon by all users, so the type is _intrinsic_ to the data. You cannot have a node without a type, though the type may be 'any'.

Whether the type is stored seems to be an implementation detail because we could have a transient shared tree that is never persisted, but still has the invariants that every node has a type and that type must be respected by all edits.

DLehenbauer replied:
@CraigMacomber - 'intrinsic schema' is now my favorite.

One reason I like it is that it applies equally to both the commonly understood "document" that starts at the root of the tree as well as to the attached augmentations, which are absent from the "document schema", but still have an intrinsic schema that writers must follow.

CraigMacomber replied:
I have renamed them to "stored" and "view". @DLehenbauer Resolve this if you are fine with these names for now.

CraigMacomber commented on packages/dds/tree/docs/wip/Container and Application Schema.md:
```diff
@@ -0,0 +1,194 @@
+interface Type {
+    name: string;
+    fields: Field[];
+    extraFields: FieldContent;
```
The purpose of the schema system is to restrict what data is allowed. While some applications might want to allow all possible values for unrecognized fields, doing so should be optional.

There are use-cases for preventing such flexible augmentations:
- It's much easier to add fields to nodes you know don't already have those fields.
- Some users may want to ensure specific nodes don't have extra data (ex: number terminals for some users)
- Its easier to properly maintain data the app does not understand if you limit it to occurring in a few explicit locations instead of literally anywhere.

There are some use-cases for allowing it anywhere as well.

What we want the defaults to be in the schema language, and what we want to encourage applications to do pattern wise is a separate issue.

At this abstraction level, there is no distinction between augmentations and other use cases for fields. I consider augmentations to be an optional design pattern which might be implemented using extraFields but could also be implemented as a field containing a list of augmentations, or an augmentations child node in a value trait.

DLehenbauer replied:
The local/global split seems to have helped, in that you can still constraint the 'V' of a Map<K, V> using 'localExtraFields' while allowing augmentation via global fields.

CraigMacomber commented on packages/dds/tree/docs/wip/Container and Application Schema.md:
```diff
@@ -0,0 +1,194 @@
+    String,
+    Boolean,
+    Serializable,
+}
```
No reason to have a node that stores a Null. There is only one possible Null value, so you can have a node with type null that stores nothing instead.

Serializable does mean Fluid's serializable. Not specifically blobs though (could be just some json, a number, a fluid handle (possibly mut also maybe not referring to a blob), json containing a flid handle.

I'll make this a bit more clear.

DLehenbauer replied:
That's a better way to do it, thanks.

CraigMacomber commented on packages/dds/tree/docs/wip/Container and Application Schema.md:
````diff
@@ -0,0 +1,194 @@
+    extraFields: FieldContent;
+    value: Value;
+}
+```
````
Thats a fine way to do it, though I'd be tempted to replace "character" with "Code Point" and `Value.string` with Value.number`. Either should work fine though.

CraigMacomber commented on packages/dds/tree/docs/wip/Container and Application Schema.md:
````diff
@@ -0,0 +1,194 @@
+}
+```
+
+These `Types` can be added as `container schema` as part of an edit op, which is considered conflicted if it tries to add a type that has a conflicting `name` and is not equal to the existing one.
````
piecemeal. As long as we require that all types used in the document exist in the schema, we can add them piecemeal, without fear of causing existing data to become out of schema as a side effect.

Commit:
move code into its own file, add docs and examples

CraigMacomber commented on packages/dds/tree/docs/wip/Container and Application Schema.md:
````diff
@@ -0,0 +1,194 @@
+function makeAnyType(name: string): Type {
+    return { name, extraFields: anyField };
+}
+```
````
FieldContent.types is an array, and thus a union.

Note that this function creates a type that allows any children. If you instead want to declare that a field permits any type, you specify its `types` as `undefined`.

If you want a field that can contain any Json value, then you use the array of all Json types. I have added Json as an example domain to help show this.

jack-williams replied:
Did you consider making union a type constructor, rather than fixing it to the structure of node field types? I could see people wanting to say that a field satisfies multiple types (also via a list). Is primarily to unblock the initial json case?

CraigMacomber replied:
makeAnyType was intended as an example to show how flexible types could be expressed.
It is no longer present in the updated code since names have been factored apart from the `Type` (which has been renamed to TreeSchema`. This allows `anyTree` to exist as a constant (which is in the code now), which replaces makeAnyType.

@jack-williams I'm a bit confused by your question. All trees currently have a type explicitly on them, and our schema system only refers to those types. There is no way (currently in this example) to apply structural constraints to children, you can only constraint what types of children are allowed. All types on fields are done via a tagged union type constructor. If you wanted to be explicit you could do:
```typescript
export type TaggedUnion = ReadonlySet<TreeSchemaIdentifier>;
export function union(
    ...types: TreeSchemaIdentifier[]
): TaggedUnion {
    return new Set(types);
}
```
I'm not sure what you mean by "a field satisfies multiple types".

Anyway, the example code has changed significantly since your comment, and there are many more comments in code now: let me know if there is still something that is unclear, or if you still have questions about this.

DLehenbauer replied:
@jack-williams - Just FYI - Craig answered my question, but I'll leave this unresolved in case you wanted to respond regarding unions.

CraigMacomber commented on packages/dds/tree/docs/wip/Container and Application Schema.md:
```diff
@@ -0,0 +1,194 @@
+
+## Use as `Container Schema`
+
+-   Document load: The application can check that their application schema match container schema, and check that the root has supported content.
```
I think the answer is "yes", but your terminology is a bit unclear:

There is a variance issue here: since we want a read/write APIs, the relation here is invariant: if the application schema is exactly the same as the container schema, then the app should be happy using the document directly (thats the case being addresses here) and it does not matter which schema is used. If the application's schema is more permissive, then there may be issues when editing, and if it is more restrictive, there may be issues when reading.

Thus I think "conforms to" should only be used to relate data to schema, not schema to schema. For schema, we should say if we mean equal to, or if one is more constrained (permits a subset of the documents) or less constrained (permits a superset of the documents)

CraigMacomber replied:
I have updated the document to refer to the example code which provides an implementation of this. Hopefully that can clear it up.

CraigMacomber commented on packages/dds/tree/docs/wip/Container and Application Schema.md:
```diff
@@ -0,0 +1,194 @@
+
+-   Document load: The application can check that their application schema match container schema, and check that the root has supported content.
+-   Writers: When inserting new content, and updating modified nodes, share-tree can check them against the schema in the document.
+    If the schema is missing, it must be added as part of the edit: for this the editing API needs to take in types which can have their desired container schema queried.
```
Yes, that is accurate.

There is the caveat that we could support ops that modify/replace a schema (either in ways were existing data will still be valid or including a process to update all existing instances).

Also someone with a partial checkout would want to download any relevant schema (or just all of them): since we want to require clients to only submit edits which are valid when they submit the edit, they need to know that there is not a schema with the same name, but different content, when they go to add a new schema. Some apps might be able to do this without checking for a schema in the document (ex: an app that centrally manages their schema and never updates schema, or one that includes hashes of schema in their name), but in general its not safe.

Commit:
fix phrasing

CraigMacomber commented on packages/dds/tree/docs/wip/Container and Application Schema.md:
```diff
@@ -0,0 +1,194 @@
+    - Wait for above to be deployed to most users.
+    - Update or configure app such that it writes the new schema to the document's schema list.
+
+If the existing data is compatible with the new schema:
```
Yep. Fixed.

jack-williams commented on packages/dds/tree/docs/wip/Container and Application Schema.md:
````diff
@@ -0,0 +1,194 @@
+    value: Value;
+}
+```
+
````
Full disclosure: the schema language I used here isknowingly inadequate. I was just exploring approaches to augmentations.

jack-williams commented on packages/dds/tree/docs/wip/Schema.ts:
```diff
@@ -0,0 +1,74 @@
+    name: string;
+    fields: Field[];
+    extraFields?: FieldContent;
+    value: Value;
```
Is `Value` a bit field, or will union types only be allowed within a field?

CraigMacomber replied:
No, it is not a bit field in this example code (but that would be a reasonable approach). This example is intentionally very minimal about what is supported functionality wise for values since how we want to handle them is still undecided (and this example isn't trying to take an opinion on that, just show an example that works): we expect them to only be used for custom primitives/terminals and not be a dimension application do anything with directly, so this makes them also likely to be an implementation detail.

Anyway, `ValueSchema.Serializable` is effectively 'any' for values, and since values do not support mutation (only replacing),  it should be straightforward for applications to have complex rules for values in specific types without needing to complicate the DDS with it (merges can't mutate values in unexpected ways like they can nodes, so values will stay in schema), so an app could just use  `ValueSchema.Serializable` and apply their own more restrictive rules in their `view schema`.

jack-williams commented on packages/dds/tree/docs/wip/Schema.ts:
```diff
@@ -0,0 +1,74 @@
+export interface Type {
+    name: string;
+    fields: Field[];
+    extraFields?: FieldContent;
```
To confirm my understanding of the semantics:
- when `extraFields` is missing then the type is a closed record. It has exactly the members specified by fields.
- when `extraFields` is present, then any field not in `fields` has the type specified by `extraFields`.

Are fields scoped? That is, do we assume that fields can have duplicate names but the type assigned to the field in the tree is the first? For example, is the following legal?
```
export const string: Type = {
    name: "foo",
    fields: [
        {
            name: "x",
            content: {
                multiplicity: Multiplicity.Value,
                types: [number],
            },
        },
        {
            name: "x",
            content: {
                multiplicity: Multiplicity.Value,
                types: [boolean],
            },
        },
    ],
    value: Value.Nothing,
};
```

CraigMacomber replied:
Yes, that is the semantics.

No, duplicate field names are not permitted. Fields are identified only by their node+key (name got renamed to key), and there is at most one field with a given key (well, there is a caveat that global and local keys are considered in separate namespaces and thus always distinct): this is true both of the types/schema and the nodes themselves. The schema now use sets which should reduce confusion over this.

The example code has been updated to explicitly address scoping, and majorly rework extraFields.

jack-williams commented:
@CraigMacomber, @DLehenbauer: I dropped some comments in - hope that is ok and hope they are useful.

Commit:
update doc

CraigMacomber commented on packages/dds/tree/docs/wip/Container and Application Schema.md:
```diff
@@ -0,0 +1,194 @@
+-   pure schema-on-read: no customization of container schema.
+-   almost pure schema-on-write: only using schema-on-read to assist with schema migration so their application schema doesn't have to deal with old formats (which unavoidably pile up in the container schema)
+    They may additionally use schema-on-read to enforce/fix some invariants the container schema system isn't expressive enough to declare (referential integrity for graph like references, numeric ranges, and any other appellation level invariants that are desired)
+
```
I have refactored this section to account for this pattern

Commit:
seperate local and global fields, and implement schema updating

Commit:
use type brands, and fix errors if found

Commit:
fix up schema examples

Commit:
update doc

Commit:
add checkCompatibility

Commit:
update some comments

Commit:
fix formatting

Commit:
more comments

Commit:
reference checkCompatibilityin document load sections

ruiterr commented:
@CraigMacomber To see whether this schema definition is versatile enough for our use cases, I tried to write a prototype of a schema converter that converts a PropertyDDS to the proposed Schema definition. If you have some time, I would appreciate if you could have a short look at the converter(https://github.com/ruiterr/FluidFramework/blob/PropertyDDSSchemaConverter/experimental/PropertyDDS/packages/property-dds/src/convertSchema.ts) to check, whether I correctly understood the schema definition.

I noticed a few aspects while trying to map the schemas:
1) How would we handle inheritance (and maybe also the related schema versioning)? The current schema definition allows me to declare a set of compatible types for a field. In principle, I could create the list of all valid types from the inheritance chain and put this into the set. Unfortunately, this would not be very extensible. The moment a new type is added that inherits from a base type, all schemas referencing the base type would have to be updated to reflect this new addition to the set of allowed types. In addition to that, it would potentially be a quite large list, which would be duplicated in many places in all schemas. I think we should introduce some mechanism to declare dependencies between the schemas.
2) Which key should I use for the default entry (in explicity sequences and strings)? At the moment, I arbitrarily choose "entries", but this risks, that every application uses a different key name, making interoperability difficult. It probably would make sense to have a default in the schema language, which indicates that this is a node that is a pure container for a sequence (and does not allow other traits).
3) Which value type should I use for the string defintion? At the moment, there is no character type. Do we need to add that?
4) Value currently only supports number. In PropertyDDS we have more fine grained number types. I assume, that we will provide pre-registered default schemas for those, which build on top of the schema language, but already specify the actual data type in a common well understood way?

Generally, it seems that most of the current PropertyDDS schema can be mapped to your proposed schema defintion (one way, it is not a bijection, many schemas could not be projected back into the PropertyDDS system).

There are a few concepts I didn't map, but these are probably on a higher level of abstraction and there is no intention to include them in this low level schema definition?
* Annotations
* Length constraints for arrays / strings
* Constants
* Values for enums
* Default values

Is there a plan where this additional information within the schema would be stored? I assume, we would need to still store this information together with the compiled schema, having something like a source schema plus a compiled schema in the document.

DLehenbauer commented on packages/dds/tree/docs/wip/JsonDomainSchema.ts:
```diff
@@ -0,0 +1,95 @@
+    extraGlobalFields: false,
+    localFields: new Map([
+        [
+            "children" as LocalFieldKey,
```
Nit: Array contents are typically referred to as 'items' rather than 'children'.

DLehenbauer commented on packages/dds/tree/docs/wip/Schema.ts:
```diff
@@ -0,0 +1,621 @@
+    /**
+     * Exactly 0 items.
+     */
+    Forbidden,
```
Do you have a use case in mind for 'Forbidden'?

The only thing that comes to mind for me is to prevent mistakes in a discriminated union (e.g., I distinguish between types 'X' and 'Y' by knowing that field 'foo' only exists on X.  For safety, I could explicitly forbid 'foo' on Y.)

DLehenbauer commented on packages/dds/tree/docs/wip/Schema.ts:
```diff
@@ -0,0 +1,621 @@
+ * An interesting alternative to this simple value Enum would be to use something more expressive here, like JsonSchema:
+ * since this is modeling immutable data, we really just need a way to figure out which if these value schema allow super sets of each-other.
+ */
+export enum ValueSchema {
```
Nit: I think it's more natural to refer to these as ValueType, FieldType, etc.

Both GraphQL and JSON schema use the term "type" for these concepts.

DLehenbauer commented on packages/dds/tree/docs/wip/Schema.ts:
```diff
@@ -0,0 +1,621 @@
+ * since this is modeling immutable data, we really just need a way to figure out which if these value schema allow super sets of each-other.
+ */
+export enum ValueSchema {
+    Nothing,
```
From context, I'm guessing 'nothing' refers to a node that has no value (i.e., it is a pure composite type).

Is this distinct from 'null'?

I assume 'undefined' exclusively means the absence of a field (as in JSON, but not JS).

CraigMacomber replied:
Yes, 'nothing' means no value.

We have no concept of 'null' at this abstraction level. Domains that with to have a concept of null can introduce one (the Json domain example does this).

This schema system does not have an opinion on what the APIs for the actual tree nodes are: they would have to decide how they implement absence of values and nothing values: in javascript, I would recommend using values which are undefined in these cases, but other options are possible.

As for the absence of a field: generally, that is not a thing: logically all fields are present, but most of them are empty. Thus you generally have an api to get the content of a field with a specific key (which might give you a 0 length sequence), an API to enumerate all non empty fields (useful for walking the tree), and APIs for accessing fields from the schema (either giving you keys to use in the first API or something more specific,).

DLehenbauer replied:
Yep, I get it.  Thx.

CraigMacomber commented on packages/dds/tree/docs/wip/Container and Application Schema.md:
```diff
@@ -0,0 +1,194 @@
+interface FieldContent {
+    multiplicity: Multiplicity;
+    types?: Type[];
+}
```
I went with FieldSchema

DLehenbauer commented on packages/dds/tree/docs/wip/Schema.ts:
```diff
@@ -0,0 +1,621 @@
+     *
+     * Usually `Multiplicity.Value` should NOT be used here since no nodes can ever be in schema are in schema if you use `Multiplicity.Value` here (that would require infinite children).
+     */
+    readonly extraLocalFields: FieldSchema;
```
Is the purpose of 'extraLocalFields' to constraint the value type when modeling a map (i.e., 'extraLocalFields' corresponds to 'V' in 'Map<K, V>')?

CraigMacomber replied:
Yes. I have updated the docs accordingly (they were pretty out of date).

DLehenbauer commented on packages/dds/tree/docs/wip/Schema.ts:
```diff
@@ -0,0 +1,621 @@
+     * If true, GlobalFieldKeys other than the ones listed above in globalFields may be used to store data on this tree node.
+     * Such fields must still be in schema with their global FieldSchema.
+     */
+    readonly extraGlobalFields: boolean;
```
I am assuming that 'augmentation' is implemented via globalFieldKeys.

If so, setting this to 'false' would prevent augmentation.  Is that something we want to support?

CraigMacomber replied:
Yes, globalFieldKeys would be my recommended pattern for implementing augmentations.

This allows applications to opt into augmentations anywhere which I think is what we want, but we could allow them everywhere unconditionally or nowhere and do augmentations a different way. I'd updated the documentation to cover this.

Commit:
call json array content items not children

Commit:
document usecases for Multiplicity.Forbidden

Commit:
Add open question about triansient out of schema states during edits

Commit:
Update docs for extraLocalFields

Commit:
document design options around TreeSchema.extraGlobalFields

CraigMacomber commented on packages/dds/tree/docs/wip/Schema.ts:
```diff
@@ -0,0 +1,621 @@
+    readonly types?: ReadonlySet<TreeSchemaIdentifier>;
+}
+
+export interface TreeSchema {
```
The name "Node" is pretty problematic due to collisions in the js ecosystem (between typescript's Node https://www.typescriptlang.org/docs/handbook/dom-manipulation.html#the-node-interface and NodeJs at least) as well as much less specific (ex: graphs have nodes). This has led to the type which these schema apply to generally being called "Tree" and not "Node", which is why I've been naming its related types Tree* an not Node*.

Commit:
Discuss bounded open polymorphism

DLehenbauer commented on packages/dds/tree/docs/wip/Container and Application Schema.md:
```diff
@@ -0,0 +1,194 @@
+    Anytime the application schema and container schema are different, this can allow the app to function as well as possible (detect and handle out of schema data, both on read and on edit, as localized as permitted by the provided error handlers).
+-   Provide at least one way to customize the container schema (schema-on-write), which can at least do type based constraints.
+-   Have at least one good design-pattern apps can use to handle schema migrations for each way they have to enforce/specify schema (make sure this supports a roll out process were clients have mixed code versions, and old documents will be supportable forever)
+
```
Yeah, it's really good.  Thank you for adding it!

Commit:
Flesh out schema migration design a bit, add example

Commit:
add notes to timline

Commit:
constants, enums and default values

CraigMacomber commented:
@ruiterr :

> 1. How would we handle inheritance (and maybe also the related schema versioning)? The current schema definition allows me to declare a set of compatible types for a field. In principle, I could create the list of all valid types from the inheritance chain and put this into the set. Unfortunately, this would not be very extensible. The moment a new type is added that inherits from a base type, all schemas referencing the base type would have to be updated to reflect this new addition to the set of allowed types. In addition to that, it would potentially be a quite large list, which would be duplicated in many places in all schemas. I think we should introduce some mechanism to declare dependencies between the schemas.

I'd assume an actual schema language (embedded DSL or otherwise) would provide a way to reuse lists of type in multiple places (we did with for whiteboard), as well as deal with cross schema dependencies (including co-recursive ones). I'm going to consider those schema language issues, which I've been trying to scope out of this document.

In response to the rest of this I have:
- Added the following sections to the doc:
   - Do we need bounded open polymorphism?
   - Approaches for bounded open polymorphism (this includes for example of explicitly listing all the types as one option, but provides some others as well)
   - Inheritance Notes
 - Updated the "Design-pattern apps can use to handle schema migrations" section to have a lot more detail, and cover how it relates to versioning.


> 2. Which key should I use for the default entry (in explicity sequences and strings)? At the moment, I arbitrarily choose "entries", but this risks, that every application uses a different key name, making interoperability difficult. It probably would make sense to have a default in the schema language, which indicates that this is a node that is a pure container for a sequence (and does not allow other traits).

I have been using "children" if there isn't something more specific that makes sense for the particular domain. As Daniel pointed out, "Items" makes sense for Json's array better than children. I think in the case where the tree is locally made entirely out of that one sequence, calling it "items" makes sense. Anyway, I don't expect this case to actually come up a lot (I expect most cases will not need an intermediate node to allow representing a difference between an empty sequence and "null"), but if it does, we should think about it more.

> 3. Which value type should I use for the string defintion? At the moment, there is no character type. Do we need to add that?

The set of value types in my example is pretty much a placeholder. If you do want to make an editiable string using them though there is an example of that in SchemaExamples.ts. I Assume we will want something smarter value wise before we commit to document compatibility (In M2), and would like to scope that out of this PR.

> 4. Value currently only supports number. In PropertyDDS we have more fine grained number types. I assume, that we will provide pre-registered default schemas for those, which build on top of the schema language, but already specify the actual data type in a common well understood way?

Yes.

> There are a few concepts I didn't map, but these are probably on a higher level of abstraction and there is no intention to include them in this low level schema definition?
> 
> * Annotations
> * Length constraints for arrays / strings
> * Constants
> * Values for enums
> * Default values
> 

TreeSchema.extraGlobalFields now has some documentation about annotations.

More length constraints (beyond what multiplicity allows), if desired, should be possible to add in the future. If we want to enforce them during merges, we probably want to handle them at this level. If we just want to check them on read, that would be easier, but both could be done, but thats more of a M2 or M3 discussion. I have added this (and some other things) to the schedule section accordingly.

I think constants, enums and default values are fine to leave for after M3. I've added some stuff about them in the schedule section. I'm not clear on what you mean by constants, but I'm guessing its ok to delay worrying about for now: if not let me know and we can discuss earlier.

> Is there a plan where this additional information within the schema would be stored? I assume, we would need to still store this information together with the compiled schema, having something like a source schema plus a compiled schema in the document.

The "Places we can store stored schema information" section covers what our options here are. My example implementation doesn't currently have an opinion on which of those we support extra data being stored in. I suspect the first/easiest place to store such data is in the source as code, and make it available on the view schema only. The ability to store extra data in the schema themselves in a way thats guaranteed to make for all apps using the document  (as part of stored schema, in some schema store, or in code with a hash check or something) could be added latter.

I don't have time today to look at your code: I'll review it and post my thoughts on it once I have time.

Commit:
capatalize FLuid

ruiterr commented:
> I'd assume an actual schema language (embedded DSL or otherwise) would provide a way to reuse lists of type in multiple places (we did with for whiteboard), as well as deal with cross schema dependencies (including co-recursive ones).

I am mainly concerned with the case, where we have schema-on-write and the schemas are stored within the document (or a central repository). For pure application specific, on-read schemas, it would certainly be possible to compile these sets of compatible types on the fly from some other schema representation (in our case, we for example could do it based on the schemas registered in the PropertyFactory). However, once the schemas are stored in the document, this dynamic compilation would no longer work. If a new type is added, the application would have to update all schemas within the document that reference this .

I assume that the documents would contain the schema using the representation you proposed here? Or do you assume it would be dynamically created from the DSL?

> Build a nominal open polymorphism system into the schema, allowing types to explicitly declare they are a member of a particular typeset/interface. Compatibility between stored and view schema can include checking these align.

That is the approach I had in mind. I was pondering to add the concept of a type-group, which can be referenced in the set of types. Schemas can be added to the type group and thus would automatically be valid everywhere, the group is referenced. I agree that the term inheritance is probably not correct here, since this is a more generic concept, since it does not specify in which way the types are related, only that all of them might appear. In contrast inheritance implies, that the types are compatible (and potentially even share code).

> I have been using "children" if there isn't something more specific that makes sense for the particular domain.

I guess it doesn't matter much, which term we use, but it should be declared in the schema.ts as a constant, so that it is used everywhere.

> Anyway, I don't expect this case to actually come up a lot (I expect most cases will not need an intermediate node to allow representing a difference between an empty sequence and "null"), but if it does, we should think about it more.

My impression was, that this is actually very common. In my schema conversion it applies to every string and every array node (you could potentially avoid it in the case, where the string/array is marked as required in the schema, but for dynamic or optional ones it is needed and thus for simplicity of implementaiton it probably should always be used).

> I'm not clear on what you mean by constants, but I'm guessing its ok to delay worrying about for now: if not let me know and we can discuss earlier.

Constants is a feature supported by the PropertyDDS schema language. It adds constant values to the data structure at runtime, which are specified in the schema. Those are not stored in the actual document. Use cases for this were things like adding a unit or documentation field. They therefore have some overlap with annotations, but they have some additional complexity, because parent schemas are allowed to overwrite the value of constants in the nested schemas (which I guess is a bit like the structural typing you mentioned). Since this does not affect the actual data in the document, it probably doesn't relate to the discussion here and I am also not sure, whether we would like to support this in the future or should instead use a different mechanism.

In the section on migration you suggest:
> -   Author new more flexible schema with same type identifier.

I am wondering, whether we should use the same type identifier or a new one (potentially using the type-group mechanism to declare compatibility). In PropertyDDS, we have always created a new type (using semver to indicate that the type is compatible), but admittedly I didn't yet consider in much detail, how a sysem with in-place updates of types would work. 

So, I'd like to more clearly understand the different pros and cons for both approaches:

*Creating a new type*
+ Many different versions of the schema can coexist in the same document and still be conformant to the schema(s)
+ In references / schemas / application code it is possible to explicitly specify which version is the minimum requirement for compatibility
+ Type identifiers are compatible between documents (i.e. two documents referring to the same version of a schema reference exactly the same data layout). This makes it easier to copy / move data between documents. Similarly, type identifiers can more easily related to external systems (e.g. the table layout in a relational database that is updated based on data from Fluid).
+ Creates a clear order in which the schemas are updated

- More tedious for application authors (they have to maintain a list of all the schema versions and add version numbers when they reference types)
- Requires a mechanism to handle bounded open polymorphism to be feasible.

*Updating existing type*
+ Simpler for application authors
+ No need to update references to the schema / use a polymorphism mechanism

- Coexistance of schema versons in one document is difficult. Either all new fields are marked as optional (in which case there could be instances in which some random subset of the optional fields is not filled, which does not correspond to the order in which the schema was updated) or the old data would be out-of-schema once the schema is updated. In the latter case we would either have to accept a schema violation or migrate all data within the document.

I think in general, the approach how we handle schema updates (having the application support the old and new schema / having a migration handler) is probably very similar in both cases.

@CraigMacomber @DLehenbauer Do you see other pros- and cons- for the different approaches?

@CraigMacomber In the design document, you mention "schematize" several times. Is this / will the be part of Fluid / the SharedTreeDDS or is this a higher level abstraction that should not be part of the design document on this level?

Commit:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into schema-doc

Commit:
reorginize code and docs into src directory

Commit:
add schema comparison tests

Commit:
add Builders.ts with schema building helpers

Commit:
add StoredSchemaRepository test/example

Commit:
add constants fro itemsKey and rootFieldKey

CraigMacomber commented:
> > I'd assume an actual schema language (embedded DSL or otherwise) would provide a way to reuse lists of type in multiple places (we did with for whiteboard), as well as deal with cross schema dependencies (including co-recursive ones).
> 
> I am mainly concerned with the case, where we have schema-on-write and the schemas are stored within the document (or a central repository). For pure application specific, on-read schemas, it would certainly be possible to compile these sets of compatible types on the fly from some other schema representation (in our case, we for example could do it based on the schemas registered in the PropertyFactory). However, once the schemas are stored in the document, this dynamic compilation would no longer work. If a new type is added, the application would have to update all schemas within the document that reference this .
> 
> I assume that the documents would contain the schema using the representation you proposed here? Or do you assume it would be dynamically created from the DSL?

Yes, my schema types are what would be stored in the document (though serialized or course: could be done by replacing the maps with objects, the sets with arrays and using JSON). See StoredSchemaRepository for details on how thats handled/updated.

I have also added StoredSchemaRepository.spec.ts which has a full executable example (as a test) going through schema changes with both stored and view schema. I think that should be easier to understand that my various descriptions of the process in the doc.

> 
> > Build a nominal open polymorphism system into the schema, allowing types to explicitly declare they are a member of a particular typeset/interface. Compatibility between stored and view schema can include checking these align.
> 
> That is the approach I had in mind. I was pondering to add the concept of a type-group, which can be referenced in the set of types. Schemas can be added to the type group and thus would automatically be valid everywhere, the group is referenced. I agree that the term inheritance is probably not correct here, since this is a more generic concept, since it does not specify in which way the types are related, only that all of them might appear. In contrast inheritance implies, that the types are compatible (and potentially even share code).

We designed, but never implemented, an approach like this for view schema in whiteboard: I think it should work well for their use cases. Adding it to the stored schema is practical as well.

If you are happy with this approach, I think we can consider bounded open polymorphism solved for now since we have at least one design we are confident we can add when it's time to implement it (prioritization for implantation and considering alternatives can happen in the M1-M2 timeframe).

> 
> > I have been using "children" if there isn't something more specific that makes sense for the particular domain.
> 
> I guess it doesn't matter much, which term we use, but it should be declared in the schema.ts as a constant, so that it is used everywhere.

Added to the new Builders.ts file along with some other handy constants.

> 
> > Anyway, I don't expect this case to actually come up a lot (I expect most cases will not need an intermediate node to allow representing a difference between an empty sequence and "null"), but if it does, we should think about it more.
> 
> My impression was, that this is actually very common. In my schema conversion it applies to every string and every array node (you could potentially avoid it in the case, where the string/array is marked as required in the schema, but for dynamic or optional ones it is needed and thus for simplicity of implementaiton it probably should always be used).

Yes, those are good places to use that pattern. I guess in practice you have a lot more of those cases than I expected.

> > I'm not clear on what you mean by constants, but I'm guessing its ok to delay worrying about for now: if not let me know and we can discuss earlier.
> 
> Constants is a feature supported by the PropertyDDS schema language. It adds constant values to the data structure at runtime, which are specified in the schema. Those are not stored in the actual document. Use cases for this were things like adding a unit or documentation field. They therefore have some overlap with annotations, but they have some additional complexity, because parent schemas are allowed to overwrite the value of constants in the nested schemas (which I guess is a bit like the structural typing you mentioned). Since this does not affect the actual data in the document, it probably doesn't relate to the discussion here and I am also not sure, whether we would like to support this in the future or should instead use a different mechanism.

Ok. We should be able to add support for that at either the view or stored schema level should use-cases arise.
I think we can lump it in with the concept of storing arbitrary extra data in the schema since it doesn't interact with data's schema compliance or merging.

> 
> In the section on migration you suggest:
> 
> > * Author new more flexible schema with same type identifier.
> 
> I am wondering, whether we should use the same type identifier or a new one (potentially using the type-group mechanism to declare compatibility). In PropertyDDS, we have always created a new type (using semver to indicate that the type is compatible), but admittedly I didn't yet consider in much detail, how a sysem with in-place updates of types would work.
> 
> So, I'd like to more clearly understand the different pros and cons for both approaches:
> 
> _Creating a new type_
> 
> Pro
> 
> * Many different versions of the schema can coexist in the same document and still be conformant to the schema(s)
> * In references / schemas / application code it is possible to explicitly specify which version is the minimum requirement for compatibility
> * Type identifiers are compatible between documents (i.e. two documents referring to the same version of a schema reference exactly the same data layout). This makes it easier to copy / move data between documents. Similarly, type identifiers can more easily related to external systems (e.g. the table layout in a relational database that is updated based on data from Fluid).
> * Creates a clear order in which the schemas are updated
> 
> Con
> 
> * More tedious for application authors (they have to maintain a list of all the schema versions and add version numbers when they reference types)
> * Requires a mechanism to handle bounded open polymorphism to be feasible.
> 
> _Updating existing type_ Pro
> 
> * Simpler for application authors
> * No need to update references to the schema / use a polymorphism mechanism
> 
> Con
> 
> * Coexistance of schema versons in one document is difficult. Either all new fields are marked as optional (in which case there could be instances in which some random subset of the optional fields is not filled, which does not correspond to the order in which the schema was updated) or the old data would be out-of-schema once the schema is updated. In the latter case we would either have to accept a schema violation or migrate all data within the document.
> 
> I think in general, the approach how we handle schema updates (having the application support the old and new schema / having a migration handler) is probably very similar in both cases.
> 
> @CraigMacomber @DLehenbauer Do you see other pros- and cons- for the different approaches?

Your list seems correct. One minor difference in how I was thinking about it:
My perspective has been that you use in place updates of schema as the mechanism to handle updates which get new types (update the parent's field's allowed types: this is always compatible with existing document content so its legal).
Your note that it requires open bounded polymorphism is a good one: in place updates are just one way to implement open bounded polymorphism. Updating type sets (as discussed above) is really a special subset of inplace updates, and could be useful here as well.

I think in general we have to support both, since they work well together:
- in place updates are can be used to prevent new type changes from bubbling up to the root, or to handle them at the root, as well as handling cases where a new field really is semantically optional, and adding new content types in existing places.
- new identifier when making incompatible changes (so data in the document is never out of schema): use an in-place update as well on its parent to allow the new type there.

> 
> @CraigMacomber In the design document, you mention "schematize" several times. Is this / will the be part of Fluid / the SharedTreeDDS or is this a higher level abstraction that should not be part of the design document on this level?

I want to develop schematize as a general purpose schema on read system for trees: it takes some kind of tree, a library for applying some schema system to that tree, and a collection of handlers for violation.
Then, as part of Fluid Tree, we can use support consumers what want to use schematize by writing support for using it with our schema system and tree type. I think it should be optional, but the encouraged way to abstract legacy schema handling, and any other schema on read use-cases.
For ease of development, it will likely initially live in a subfolder inside of Fluid Tree (after its ported out if whiteboard), but someday might get its own package.

Commit:
add support for adapters

Commit:
fix tree schema comparison bug, and misc cleanup

Commit:
Rename Multiplicity to FieldKind and adjust its meaning

DLehenbauer reviewed (approved):
(@ruiterr, @jack-williams - I'm approving the PR, but still feel free to continue the dialog.)

Commit:
captitalize Fluid

Commit:
Deleate unused file

Commit:
Merge branch 'main' into schema-doc

Commit:
fix build from merge

Commit:
Add copyright
